src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -639,3 +640,151 @@ int32_t CryptoNative_SslGetCurrentCipherId(SSL* ssl, int32_t* cipherId)
+
+        ASN1_TIME_set(time, 0);
+        X509_set1_notBefore(cert, time);
+        X509_set1_notAfter(cert, time);
+
+        ret = X509_sign(cert, evp, EVP_sha256());
+    }
+
+    if (bn != NULL)
+    {
+        BN_free(bn);
+    }
+
+    if (rsa != NULL)
+    {
+        RSA_free(rsa);
+    }
+
+    if (time != NULL)
+    {
+        ASN1_TIME_free(time);
+    }
+
+    return ret;
+}
+
+int32_t CryptoNative_OpenSslGetProtocolSupport(SslProtocols protocol)
+{
+    int ret = 0;
+
+    SSL_CTX* clientCtx = CryptoNative_SslCtxCreate(TLS_method());
+    SSL_CTX* serverCtx = CryptoNative_SslCtxCreate(TLS_method());
+    X509 * cert = X509_new();
+    EVP_PKEY* evp = CryptoNative_EvpPkeyCreate();
+    BIO *bio1 = BIO_new(BIO_s_mem());
+    BIO *bio2 = BIO_new(BIO_s_mem());
+
+    SSL* client = NULL;
+    SSL* server = NULL;
+
+    if (clientCtx != NULL && serverCtx != NULL && cert != NULL && evp != NULL && bio1 != NULL && bio2 != NULL)
+    {
+        CryptoNative_SetProtocolOptions(serverCtx, protocol);
+        CryptoNative_SetProtocolOptions(clientCtx, protocol);
+        SSL_CTX_set_verify(clientCtx, SSL_VERIFY_NONE, NULL);
+        SSL_CTX_set_verify(serverCtx, SSL_VERIFY_NONE, NULL);
+
+        if (MakeSelfSignedCertificate(cert, evp))
+        {
+            CryptoNative_SslCtxUseCertificate(serverCtx, cert);
+            CryptoNative_SslCtxUsePrivateKey(serverCtx, evp);
+
+            server = CryptoNative_SslCreate(serverCtx);
+            SSL_set_accept_state(server);
+
+            client = CryptoNative_SslCreate(clientCtx);
+            SSL_set_connect_state(client);
+
+            // set BIOs in opposite
+            SSL_set_bio(client, bio1, bio2);
+            SSL_set_bio(server, bio2, bio1);
+            // SSL_set_bio takes ownership so we need to up reference since same BIO is shared.
+            BIO_up_ref(bio1);
+            BIO_up_ref(bio2);
+            bio1 = NULL;
+            bio2 = NULL;
+
+            // Try handshake, client side first.
+            SSL* side = client;
+            int sslError = 0;
+            while (1)
+            {
+                ret = SSL_do_handshake(side);
+                if (ret == 1)
+                {
+                    break;
+                }
+
+                sslError = SSL_get_error(side, ret);
+                if (sslError != SSL_ERROR_WANT_READ)
+                {
+                    break;
+                }
+
+                side = side == client ? server : client;
+            }
+        }
+    }
+
+    if (cert != NULL)
+    {
+        X509_free(cert);
+    }
+
+    if (evp != NULL)
+    {
+        CryptoNative_EvpPkeyDestroy(evp);
+    }
+
+    if (bio1)
+    {
+        BIO_free(bio1);
+    }
+
+    if (bio2)
+    {
+        BIO_free(bio2);
+    }
+
+    if (client != NULL)
+    {
+        SSL_free(client);
+    }
+
+    if (server != NULL)
+    {
+        SSL_free(server);
+    }
+
+    return ret;
```

Comment:
It's a little weird that you return whatever the last ret value was... is this guaranteed to be 1 or 0? Seems like it's clearer to just return 1 or 0 to indicate success/failure

Reply:
updated. OpenSSL uses 1 for success but I agree we do not need to bring it up.